// pusher-platform/lib/app.js
"use strict";
var extend = require("extend");
var jwt = require("jsonwebtoken");
var authenticator_1 = require("./authenticator");
var base_client_1 = require("./base_client");
var App = (function () {
    function App(options) {
        this.appId = options.appId;
        var keyParts = options.appKey.match(/^([^:]+):(.+)$/);
        if (!keyParts) {
            throw new Error("Invalid app key");
        }
        this.appKeyId = keyParts[1];
        this.appKeySecret = keyParts[2];
        this.client = options.client || new base_client_1.default({
            host: options.cluster,
        });
        this.authenticator = new authenticator_1.default(this.appId, this.appKeyId, this.appKeySecret);
    }
    App.prototype.request = function (options) {
        options = this.scopeRequestOptions("apps", options);
        if (options.jwt == null) {
            options = extend(options, { jwt: this.generateSuperuserJWT() });
        }
        return this.client.request(options);
    };
    App.prototype.configRequest = function (options) {
        options = this.scopeRequestOptions("config/apps", options);
        if (options.jwt == null) {
            options = extend(options, { jwt: this.generateSuperuserJWT() });
        }
        return this.client.request(options);
    };
    App.prototype.authenticate = function (request, response, options) {
        this.authenticator.authenticate(request, response, options);
    };
    App.prototype.scopeRequestOptions = function (prefix, options) {
        var path = ("/" + prefix + "/" + this.appId + "/" + options.path)
            .replace(/\/+/g, "/")
            .replace(/\/+$/, "");
        return extend(options, { path: path });
    };
    App.prototype.generateSuperuserJWT = function () {
        var now = Math.floor(Date.now() / 1000);
        var claims = {
            app: this.appId,
            iss: this.appKeyId,
            su: true,
            iat: now - 30,
            exp: now + 60 * 5,
        };
        return jwt.sign(claims, this.appKeySecret);
    };
    return App;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = App;
